Open Policy Agent support

spacedub opened the pull request:
This is a tentative implementation providing OPA support (fix #373 ).

Here is how to test.
Create a policy.rego file:
```
package policy

import input.certificate

default allow := false

allow = true {
	input.certificate.Subject.CommonName == "XXXX Whatever is in your cert"
}
```

Call ghost:
```
./ghostunnel server [...] --allow-policy policy.rego --allow-query "data.policy.allow"
```

Problems with this PR:
* [x] client support is not implemented yet
* [x] I am not familiar enough with OPA to be sure of what I am doing checking the result - specifically, should I iterate over `results[0].Expressions`?
* [x] for some reasons, adding the dependency to OPA with go mod also upgrades a lot of other libraries - not quite sure why, or what I should do instead
* [x] UX is tentative - comments?
* [x] I am certainly not the best go developer - anything I should fix here?
* [x] no tests
* [x] no documentation

Ping @csstaub

csstaub commented:
Regarding libraries: Go uses "minimum required dependency" (as opposed to maximum like many other languages) for resolving versions. Most likely OPA or its transitive dependencies has higher minimums, so that bumps up the versions. That's totally fine.

Regarding results: Yes, I believe it's possible for the policy to return multiple results, but we should probably just standardize on the convention that the result must be a specific variable (`allow`) that is of type bool. 

Regarding the UX: The `--allow-policy` and `--allow-query` flags make a lot of sense to me, I think that's good. And we should make them mutually exclusive with other flags.

csstaub commented on auth/auth.go:
```diff
@@ -101,6 +107,33 @@ func (a ACL) VerifyPeerCertificateServer(rawCerts [][]byte, verifiedChains [][]*
+
+	// Check against OPA
+	if a.AllowOPAQuery != nil {
+		results, err := a.AllowOPAQuery.Eval(context.TODO(), rego.EvalInput(input))
```
Let's thread this context from the proxy module. In the proxy, each new connection should create a context and then pass it along to here. This way, we can set a deadline for new connections in the proxy module in the future to avoid hanging if a policy takes too long to execute.

csstaub replied:
A good default timeout for querying the policy is probably around 1s, what do you think?

csstaub replied:
Maybe we could re-use the connect timeout, but I just realized threading the timeout through might be tricky because this is in a callback. So maybe instead of threading a context through we can just put the connect timeout on the eval directly

spacedub replied:
Ok - makes sense. I will wire a timeout on the eval then.

csstaub commented on auth/auth.go:
```diff
@@ -101,6 +107,33 @@ func (a ACL) VerifyPeerCertificateServer(rawCerts [][]byte, verifiedChains [][]*
+		results, err := a.AllowOPAQuery.Eval(context.TODO(), rego.EvalInput(input))
+		// rst, _ := json.MarshalIndent(results, "", "  ")
+		// fmt.Printf("%s\n", rst)
+		if err != nil {
```
Let's use `rego.ResultSet.Allowed()` here, It will return true/false based on OPA's usual conventions:
https://pkg.go.dev/github.com/open-policy-agent/opa/rego#ResultSet.Allowed

I think this should work:
```
if err != nil {
  // error handling
}

if results.Allowed() {
  return nil
}
```

csstaub commented on auth/auth.go:
```diff
@@ -101,6 +107,33 @@ func (a ACL) VerifyPeerCertificateServer(rawCerts [][]byte, verifiedChains [][]*
 	}
 
+	input := map[string]interface{}{
+		"certificate": cert,
```
What does the input into the policy look like if we provide the cert like this? Any pre-processing we should do in order to make it easier to write a policy against? I'm wondering what extensions would look like in the input, for example.

spacedub replied:
Yep, not sure what happens with extensions. For basic stuff, it is just the x509 go object reflected. eg: `input.certificate.Subject.CommonName`

Will check for extensions.

I would have a slight preference for *not* doing any pre-processing / forcing into a custom API - and instead just letting people access the certificate and do whatever they want.

By the way, I am wondering if we should also expose other connection information to the policy, besides the certificate (client ip?)?

csstaub replied:
I prefer less processing too, but at least we should make sure we have DNS and URI SANs easily available because that's how modern standard certificates are issued for browsers. Exposing other info sounds useful, esp. the IP address of the client could be interesting.

spacedub replied:
Agreed on making sure we can access these  - will check.

About the IP - after checking quickly: might be tricky to get the IP - since we do not have access to the connection inside `VerifyPeerCertificateClient` - any suggestion on how to best pass it along here?

csstaub commented on main.go:
```diff
@@ -536,14 +552,26 @@ func serverListen(context *Context) error {
+		rego.Query(*serverAllowQuery),
+		rego.Load([]string{*serverAllowPolicy}, nil),
+		//rego.Module(ct.String("policy"), string(module)),
+	).PrepareForEval(gocontext.TODO())
```
This can be `gocontext.Background()`, since we're doing this processing at startup and don't intent to put a deadline/cancel func on it.

spacedub commented:
Updated PR:
* rebased
* added client support
* addressed some comments, specifically:
  * `Eval` now has a 1s context timeout
  * OPA eval is now checked against `rego.ResultSet.Allowed()`
  * `PrepareForEval` is now with `context.Background`
 
Two pending comments:
 * verify what happens for extensions
 * consider adding connection properties available for OPA to check (clientIP) - how-to is TBD

spacedub commented on auth/auth_test.go:
```diff
@@ -43,6 +45,88 @@ var fakeChains = [][]*x509.Certificate{
+	assert.Nil(t, testACL.VerifyPeerCertificateServer(nil, fakeChains), "Rego policy validates CN should pass")
+}
+
+func TestOPAAcceptDNSn(t *testing.T) {
```
cc @csstaub 
See this test and below.

spacedub commented:
@csstaub 
Updated PR, with more tests.
Is this getting close?

Out of the list, the only things that are missing RN are:
1. documentation
2. possibly expending the scope of this beyond purely peer certificate validation and letting the OPA policy evaluate other aspects of the connection (IP? TLS version? Cipher suites? SNI?)

I would suggest we punt on 2. for now, and that could be part of subsequent work if someone is interested - that would likely require substantial changes though (as the cert verification hook does not provide the required info)

Let me know your thoughts about all this - and your opinion on the readiness of the PR.

If we are close, happy to push it through the finish line today (with some documentation and whatever small changes required).

csstaub commented on auth/auth.go:
```diff
@@ -101,6 +108,23 @@ func (a ACL) VerifyPeerCertificateServer(rawCerts [][]byte, verifiedChains [][]*
+		results, err := a.AllowOPAQuery.Eval(ctx, rego.EvalInput(input))
+		if err != nil {
+			// TODO logging? Or?
+			return err
```
Let's just return the error here, no logging. But we should wrap it:
```
return fmt.Errorf("unauthorized: policy returned error: %w", err)
```

Same thing below for client

csstaub commented on auth/auth.go:
```diff
@@ -101,6 +108,23 @@ func (a ACL) VerifyPeerCertificateServer(rawCerts [][]byte, verifiedChains [][]*
 
+	// Check against OPA
+	if a.AllowOPAQuery != nil {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```
We can address this separately, but I think the timeout should probably be configurable (either use the connect timeout or provide a separate flag for it). But let's stick with 1s for now.

spacedub replied:
Adding a note about that in the list of caveats for now

csstaub commented on auth/auth.go:
```diff
@@ -147,6 +171,23 @@ func (a ACL) VerifyPeerCertificateClient(rawCerts [][]byte, verifiedChains [][]*
+		results, err := a.AllowOPAQuery.Eval(ctx, rego.EvalInput(input))
+		if err != nil {
+			// TODO logging? Or?
+			return err
```
Same as above, wrap error here.

csstaub reviewed (approved):
I think this looks great, thanks @spacedub! I left a couple of comments but we can merge this and then continue development in separate pull requests if you want.

spacedub commented:
Thanks for all the help @csstaub 
Really appreciate the guidance.

About integration tests - are they failing because of this changeset or is there something wrong with them overall?

spacedub commented:
Your last comments have been addressed @csstaub 
Would love to see this merged at your convenience :-) (if the issue with integration tests is unrelated)

csstaub commented:
Looks like the integration test failures are related:
```
[4636] 2022/09/14 15:30:05.476954 starting ghostunnel in server mode
[921](https://github.com/ghostunnel/ghostunnel/actions/runs/3050198123/jobs/4925473709#step:3:922)
[4636] 2022/09/14 15:30:05.477083 using keystore file on disk as certificate source
[922](https://github.com/ghostunnel/ghostunnel/actions/runs/3050198123/jobs/4925473709#step:3:923)
[4636] 2022/09/14 15:30:05.480381 using target address unix:/tmp/tmpe3tbb1ky/ghostunnel-status-socket
[923](https://github.com/ghostunnel/ghostunnel/actions/runs/3050198123/jobs/4925473709#step:3:924)
[4636] 2022/09/14 15:30:05.480433 Invalid rego policy or query: cannot evaluate empty query
[924](https://github.com/ghostunnel/ghostunnel/actions/runs/3050198123/jobs/4925473709#step:3:925)
[4636] 2022/09/14 15:30:05.480439 error from server listen: cannot evaluate empty query
[925](https://github.com/ghostunnel/ghostunnel/actions/runs/3050198123/jobs/4925473709#step:3:926)
--- FAIL: TestIntegrationMain (0.00s)
[926](https://github.com/ghostunnel/ghostunnel/actions/runs/3050198123/jobs/4925473709#step:3:927)
    main_test.go:72: got error from run: cannot evaluate empty query
```

Looks like it's trying to eval a query, even when there's no policy set?

csstaub commented on main.go:
```diff
@@ -536,14 +552,26 @@ func serverListen(context *Context) error {
+	allowQuery, err := rego.New(
+		rego.Query(*serverAllowQuery),
+		rego.Load([]string{*serverAllowPolicy}, nil),
+	).PrepareForEval(gocontext.Background())
```
Looks like this block needs to be gated by the flag, so we only compile the query if both the flags for policy/query were set.

spacedub replied:
Geez
Fixing now

spacedub replied:
Adding integration tests while I am at it to avoid further idiocy :/

spacedub commented:
@csstaub latest change-set:
- fixed non functioning client mode
- gated policy eval to actual flags... 🤦 
- existing integration tests pass locally (with go1.18)
- added integration tests for both client and server OPA mode, confirming this at least "work"

Apologies for the churn here - should have been more thorough at the first try.

csstaub commented:
Tests all passed, merged into the main branch now. Thanks @spacedub
